### PR TITLE
Revamp how the cockroachdb/cockroach image is built.

### LINKED ISSUE #1754: ci: revamp how the "cockroachdb/cockroach" image is built
Currently this image is built using the `build/build-docker-deploy.sh` script which requires a "cockroachdb/cockroach-dev" image. This is massively inefficient with the new `build/builder.sh` CI build system. Instead, `build-docker-deploy.sh` should be doing something like:

```
  build/builder.sh make testbuild
  build/builder.sh make STATIC=1 release
```

And then use the resulting binary and `*.test` files to construct the docker image.

### PR DESCRIPTION
**Pull request by petermattis**
We now use build/builder.sh for building the cockroach binary and tests
and run the new acceptance tests instead of run/local-cluster.sh. The
latter is now deleted.

Fixes #1754.

### CODE REVIEW

File: build/deploy/mkimage.sh [hunk 1]
```diff
@@ -12,26 +12,35 @@
...
-cd -P "$(dirname $0)"
-DIR=$(pwd -P)
+# This is mildly trick: This script runs itself recursively. The first
+# time it is run it does not take the if-branch below and executes on
+# the host computer. It uses the builder.sh script to run itself
+# inside of docker passing "docker" as the argument causing the
+# commands in the if-branch to be executed within the docker
+# container.
+if [ "${1:-}" = "docker" ]; then
+    time make testbuild
```
Comment (petermattis) [inline build/deploy/mkimage.sh]: Is building the tests statically like this worthwhile? It takes quite a while and seems duplicative of the tests that are run build `circle-test.sh`.

File: build/deploy/cockroach.sh [hunk 2]
```diff
@@ -2,13 +2,13 @@
...
-if [ -f ./test.sh ]; then
-  ./test.sh
+if [ -f ../build/test.sh ]; then
+  ../build/test.sh
 fi
 
 if [ "${1:-}" = "shell" ]; then
   shift
-  /bin/bash "$@"
+  /bin/sh "$@"
```
Comment (bdarnell) [inline build/deploy/cockroach.sh]: Why change this to `sh`? We use `bash` in all our `#!` lines and use bashisms, well, unabashedly.

File: build/build-docker-deploy.sh [hunk 3]
```diff
@@ -2,7 +2,4 @@
 
 set -eu
 
-cd "$(dirname $0)/.."
-
-./build/build-docker-dev.sh
-./build/deploy/mkimage.sh
+$(dirname $0)/deploy/mkimage.sh
```
Comment (bdarnell) [inline build/build-docker-deploy.sh]: Why keep this file around if it's just a wrapper around mkimage.sh?

File: build/deploy/cockroach.sh (hunk 2 above)
Comment (petermattis) [inline build/deploy/cockroach.sh]: Oops, I was thinking of switching to use `busybox` as the base image, but reverted that. I'll revert this as well.

File: build/deploy/mkimage.sh [hunk 4]
```diff
@@ -12,26 +12,35 @@
 
 set -euo pipefail
 
-cd -P "$(dirname $0)"
-DIR=$(pwd -P)
+# This is mildly trick: This script runs itself recursively. The first
```
Comment (bdarnell) [inline build/deploy/mkimage.sh]: s/trick/tricky/

File: build/build-docker-deploy.sh (hunk 3 above)
Comment (petermattis) [inline build/build-docker-deploy.sh]: It's mentioned in the docs? Perhaps I should move the contents of `deploy/mkimage.sh` to this file.

File: build/deploy/mkimage.sh (hunk 4 above)
Comment (petermattis) [inline build/deploy/mkimage.sh]: Fixed.

File: build/deploy/mkimage.sh (hunk 1 above)
Comment (bdarnell) [inline build/deploy/mkimage.sh]: I don't think so; even without the build time there's a non-trivial amount of complexity devoted to getting these tests built and running in a strange artificial environment.

**Comment by bdarnell** [comment]
LGTM

File: build/build-docker-deploy.sh (hunk 3 above)
Comment (petermattis) [inline build/build-docker-deploy.sh]: This is done. I've moved the contents of `deploy/mkimage.sh` into `build/build-docker-deploy.sh`.

**Comment by petermattis** [comment]
PTAL at the new commit.

**Comment by bdarnell** [comment]
LGTM

**Comment by tschottdorf** [comment]
the reason we were running the tests again in the deploy image was that the static build can behave differently. For example, failure to correctly use `netgo` and putting the "static" binary into a `busybox` will cause all DNS lookups to fail. I don't think this warrants all the effort put into it previously, so as long as there's some acceptance test that makes sure the deploy image isn't completely toast, that'll do.

**Comment by petermattis** [comment]
@tschottdorf Yep, I'm still running the acceptance tests. Curious about the '*.test' binaries, though. They weren't being built with `-tags netgo`, so the DNS behavior would be different from the cockroach binary.

**Pull request merged**